presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedHistogram.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.block.VariableWidthBlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.type.TypeUtils;
+import com.facebook.presto.util.array.IntBigArray;
+import com.facebook.presto.util.array.LongBigArray;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.type.TypeUtils.buildStructuralSlice;
+import static com.facebook.presto.type.TypeUtils.expectedValueSize;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static it.unimi.dsi.fastutil.HashCommon.arraySize;
+import static it.unimi.dsi.fastutil.HashCommon.murmurHash3;
+
+public class TypedHistogram
+{
+    private static final float FILL_RATIO = 0.9f;
+
+    private int maxFill;
+    private int mask;
+
+    private Type type;
+
+    // the hash table from values to groupIds
+    private BlockBuilder values;
```

Comment:
Do you need all these fields? It seems like `values` and `valuesByGroupId` are the same thing, `groupIds[i]` is `hashPositions[i] - 1` and `nextGroupId` == `values.getPositionCount`